### PR TITLE
Ensure an alias is set when creating a join

### LINKED ISSUE #28724: [MLv2] Make sure adding joins generates nice join aliases and they get included in the `:join-alias` for joined Fields
Adding a join works like, or will work like, something like this:

1. I have a query against Table A
2. I fetch a list of Tables I can join (this will be implemented in https://github.com/metabase/metabase/issues/28699)
3. I pick a Table, B
4. I fetch a list of Columns (presumably in Table A) I can use as the LHS of the join's condition, `=` by default. (This will be implemented in https://github.com/metabase/metabase/issues/28703)
5. I pick a column for the LHS
6. I fetch a list of Columns (presumably in Table B) I can use as the RHS of the condition
7. I pick a column for the RHS
8. I add that join + condition to the query (already implemented)

JS code might look something like

```js
let query = Query.forTable(tableA);
const joinableTables = query.joinableTables();
// display joinableTables and user picks one ...
const tableToJoin = joinableTables[1];
const lhsColumns = query.filterableFields("=");
// display lhsColumns and user picks one ...
const lhsColumn = lhsColumns[0];
const rhsColumns = query.filterableFields(tableToJoin, "="); // not 100% sure about how this should look
// display rhsColumns and user picks one ...
const rhsColumn = rhsColumns[3];

// now add the join to the query
query = query.join(tableToJoin, equals(lhsColumn, rhsColumn));
```

This is all cool, but we need to make sure that `:field` clauses in the join condition that come from Table B  include the generated `:join-alias` in their options. Currently the do not. We need to look at the `:lib/stage-metadata` and also the metadata for whatever we join to determine this info.

This might be a little tricky because we'll have to come up with the join alias when we add the join, and that will have to happen before we resolve the condition... it will all have to be done as we add the join to the query I guess. We probably need to do https://github.com/metabase/metabase/issues/28719 first before tackling this

### PR DESCRIPTION
**Pull request by metamben**
Fixes #28724.

### CODE REVIEW

File: test/metabase/lib/join_test.cljc [hunk 1]
```diff
@@ -74,28 +86,36 @@
...
           (is (=? {:lib/type    :mbql/join
                    :lib/options {:lib/uuid string?}
                    :stages      [{:lib/type     :mbql.stage/mbql
                                   :source-table (meta/id :venues)}]
                    :conditions  [[:=
                                   {:lib/uuid string?}
-                                  [:field {:lib/uuid string?} (meta/id :venues :category-id)]
-                                  [:field {:base-type :type/BigInteger, :lib/uuid string?} "ID"]]]}
+                                  [:field {:base-type :type/BigInteger, :lib/uuid string?} "ID"]
+                                  [:field {:lib/uuid string?} (meta/id :venues :category-id)]]]}
```
Comment (metamben) [inline test/metabase/lib/join_test.cljc:98]: Such fuzzy cases only work if the LHS contains the home column and the RHS the joined column, so I had to swap the references.

File: src/metabase/lib/join.cljc [hunk 2]
```diff
@@ -271,6 +277,117 @@
...
+        ;; otherwise the first PK home column
+        (m/find-first lib.types.isa/primary-key? home-cols)
+        ;; otherwise the first home column
+        (first home-cols))))
+
+(defn- strip-id [s]
+  (when (string? s)
+    (str/trim (str/replace s #"(?i) id$" ""))))
+
+(defn- similar-names [name0 name1]
```
Comment (camsaul) [inline src/metabase/lib/join.cljc:305]: I'm sure this does something important but I am really having a tough time understanding what it does just based on the name. Is this a predicate function... in which case should it be called `similar-names?`. Even tho this is private and we don't require a docstring, a short docstring would go a long way in making this code easier to grok

File: src/metabase/lib/join.cljc [hunk 3]
```diff
@@ -271,6 +277,117 @@
...
+(comment
+  (similar-names "jellow" "yellow")
+  (similar-names "yellow" "yellow")
+  (similar-names "yellow" "yellows")
+  (similar-names "yellow" "yellows")
+  (similar-names "yellows" "yellows")
+  (similar-names "cactus" "cacti")
+  (similar-names "cactus" "cactuses")
+  (similar-names "emacs" "emacsen")
+  nil)
```
Comment (camsaul) [inline src/metabase/lib/join.cljc:323]: Maybe you can take all of these test cases and stick them in a unit test for `similar-names`?

File: src/metabase/lib/join.cljc (hunk 2 above)
Comment (metamben) [reply src/metabase/lib/join.cljc:305]: Added the question mark and a comment. (BTW, this is basically ported from the FE code.)

File: src/metabase/lib/join.cljc (hunk 3 above)
Comment (metamben) [reply src/metabase/lib/join.cljc:323]: They are not all valid, I was just experimenting. Removed them.

**Review by camsaul — APPROVED** [review]

**Pull request merged**